packages/pwa-kit-react-sdk/src/ssr/universal/hooks/use-server-effect.tsx

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import React, {useContext, useState, useEffect} from 'react'
+import {useUID} from 'react-uid' // TODO: Probably not required
+import useGetPropsArgs, {GetPropsArgs} from './use-get-props-args'
+
+// Globals
+export const DEFAULT_CONTEXT_KEY = '__SERVER_EFFECTS__'
+const contexts = []
+const isServer = typeof window === 'undefined'
+
+// Type Definitions
+type ServerEffectState = {
+    data: object;
+    loading: boolean;
+    error: Error;
+}
+type DidUpdateFn = (serverEffectArgs: GetPropsArgs) => Promise<any>
+type EffectsResolver = () => Promise<{[x: string]: {}}>
+type Provider = ({ children, value }: { children: any; value: any; }) => JSX.Element
+type ServerEffect = (initialValue: Object, didUpdate: DidUpdateFn, source: any[]) => ServerEffectState
+type ServerEffectContext = {
+    /**
+    * ServerEffectProvider: The React provider you must wrap you application with in order to use
+    * the accompanied hook.
+    */
+    ServerEffectProvider: Provider; // Can we get more specifc here? provider type?
+    /**
+    * useServerEffect: A useEffect style hook that can be used client and server side to run async
+    * actions.
+    */
+    useServerEffect: ServerEffect;
+    /**
+    * resolveEffects: For use on the server side only, this async function will can be called to invoke
+    * all the recorded async "didUpdate" functions.
+    */
+    resolveEffects: EffectsResolver;
+}
+type ServerEffectContextValue = {
+    name: string;
+    data: object;
+    requests: DidUpdateFn[];
+    resolved: boolean;
+}
+
+/**
+ * This function is creating `useServerEffect`, it's only going to be used
+ * but the SDK, this isn't for external libs.
+ * 
+ * @param {*} context 
+ * @returns 
+ */
+const createServerEffect = (context) => {
+    const useServerEffect = (didUpdate: DidUpdateFn, source: any[], initial?: any) : ServerEffectState => {  
+        debugger  
+        const key = useUID()
+        const ctx: ServerEffectContextValue = useContext(context)
```

Comment:
```suggestion
        const ctx = useContext(context)
```
If the types are properly defined elsewhere, then TypeScript should be able to infer this type.